Add RelationalDbSet for raw query

mikary opened the pull request:
Add custom relational queries off of DbSet to provide functionality for #624
@ajcvickers  @anpete @divega

anpete commented on src/EntityFramework.Core/Utilities/ServiceProviderExtensions.cs:
```diff
@@ -11,7 +11,7 @@
 namespace Microsoft.Framework.DependencyInjection
 {
-    internal static class ServiceProviderExtensions
+    public static class ServiceProviderExtensions
```
If you need these, share them, rather than making this public.

anpete commented:
Can you please describe what's in this change?

anpete commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -148,5 +158,82 @@ var keyFactory
                     _readerParameter));
         }
+
+        protected Expression VisitRelationalCustomQueryable(string sql, Type elementType)
```
Don't duplicate the code below. Refactor both into something that pivots on the source table expr.

mikary replied:
Would we want an optional expression parameter so the existing code path can use the element type as it was but we still have a way to pass in the query string?

anpete replied:
No, just extract a new method and call it from both places. It may be DRY'er to have the new method take a Func<TableExpressionBase> as a factory for the table expr. but I would see what it looks like to just build the table expr. externally first.

anpete commented on src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs:
```diff
@@ -163,6 +163,23 @@ public virtual Expression VisitSelectExpression(SelectExpression selectExpressio
+        {
+            Check.NotNull(customTableExpression, nameof(customTableExpression));
+
+            if (customTableExpression.Schema != null)
```
Don't think this is needed, is it?

anpete commented on src/EntityFramework.Relational/Query/Expressions/CustomTableExpression.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public virtual string Table { get; }
+
+        public virtual string Schema { get; }
```
Don't think this is needed, is it?

mikary replied:
This is how the ```DefaultSqlQueryGenerator``` is able to distinguish a table from a raw query so it knows to make a sub query.

anpete replied:
Meant Schema.

anpete commented on src/EntityFramework.Relational/Query/Expressions/CustomTableExpression.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.Data.Entity.Relational.Query.Expressions
+{
+    public class CustomTableExpression : TableExpressionBase
```
Can we rename this to "RawSqlDerivedTableExpression"?

mikary replied:
"Custom" is a pretty bad name, it would probably be better as "RawSql" in most places

anpete commented on src/EntityFramework.Relational/Query/RelationalCustomQueryProvider.cs:
```diff
@@ -0,0 +1,95 @@
+
+namespace Microsoft.Data.Entity.Relational.Query
+{
+    class RelationalCustomQueryProvider : IAsyncQueryProvider
```
Let's try and do this using a custom LINQ operator instead. Take a look at how AsNoTracking and Include are inplemented. The difference here is that the op will be defined in Relational and so we will need to figure out how it gets registered - currently this happens in CompiledQueryCache.CreateNodeTypeProvider

anpete replied:
Ideas for the operator name:

1) "FromSql"

cc @divega

divega replied:
I had the same in mind. +1

anpete commented on src/EntityFramework.Relational/RelationalDbSetExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Microsoft.Data.Entity
+{
+    public static class RelationalDbSetExtensions
```
This can go away and the new operator can just hang from IQueryable<T>

divega replied:
Does this comment still apply? Sounds to me like we actually only want this to be callable on a root, i.e. on a DbSet<T> by I might be missing something.

mikary replied:
I don't think this is applicable anymore. The operator approach resulted in unwanted client side processing and leaves questions about how to process custom SQL at an arbitrary node in the query.

anpete commented on test/EntityFramework.Relational.FunctionalTests/QueryRelationalTest.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+        }
+
+        protected const string customQueryableSimple = "SELECT * FROM Customers";
```
Inline these. Optimize for readability in tests.

mikary commented:
Updated based on PR feedback

maumar commented on src/EntityFramework.Core/Query/EntityQueryable`.cs:
```diff
@@ -28,5 +34,46 @@ IAsyncEnumerator<TResult> IAsyncEnumerable<TResult>.GetEnumerator()
+
+        public override string ToString()
+        {
+            return base.ToString() + string.Join(", ", _annotatable.Value.Annotations.Select(annotation => annotation.Value));
```
fix the formatting - e.g. space between base.ToString() and annotations, maybe wrap annotations in some kind of parenthesis.

maumar commented on src/EntityFramework.Core/Query/EntityQueryable`.cs:
```diff
@@ -28,5 +34,46 @@ IAsyncEnumerator<TResult> IAsyncEnumerable<TResult>.GetEnumerator()
+            return base.ToString() + string.Join(", ", _annotatable.Value.Annotations.Select(annotation => annotation.Value));
+        }
+
+        private class Annotatable : MetadataBase
```
Consider more explicit name - not super important since it's private class

anpete replied:
I think we should just change MetadataBase to be this and move it out of metadata.

mikary replied:
I like the idea of having a non-abstract version of this class for re-usability (Trying to keep a DRY solution where we have to avoid multiple-inheritance)
Any thoughts on this thread? @bricelam

bricelam replied:
I'm fine with either inheritance or composition for sharing the implementation. If everything can inherit then yes, we should rename MetadataBase to Annotatable and move it to Infrastructure.

mikary replied:
Added an issue (#1778) to track this, since it's not directly related to this PR.

maumar commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -68,18 +69,25 @@ protected override Expression VisitMethodCallExpression([NotNull] MethodCallExpr
+        protected override Expression VisitConstantExpression(ConstantExpression constantExpression)
         {
-            Check.NotNull(elementType, nameof(elementType));
+            var entityQueryable = constantExpression.Value as IEntityQueryable;
```
missing check.notnull

maumar commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -68,18 +69,25 @@ protected override Expression VisitMethodCallExpression([NotNull] MethodCallExpr
 
-            var queryMethodInfo = RelationalQueryModelVisitor.CreateValueReaderMethodInfo;
-            var entityType = QueryModelVisitor.QueryCompilationContext.Model.GetEntityType(elementType);
+            if (entityQueryable != null)
```
collapse into one condition

maumar commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -87,6 +95,33 @@ protected override Expression VisitEntityQueryable(Type elementType)
+            Type elementType,
+            Func<IEntityType, string, TableExpressionBase> createTableExpression)
+        {
+            Check.NotNull(elementType, nameof(elementType));
```
i would move the check to the public surface and remove it from here

maumar commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -87,6 +95,33 @@ protected override Expression VisitEntityQueryable(Type elementType)
+            return VisitSelectExpression(elementType,
+                (entityType, tableName) =>
+                    new RawSqlDerivedTableExpression(
+                        rawSql,
```
consider not passing rawsql as closure to avoid extra string allocation (especially that rawsql can be arbitrarily large)

mikary replied:
This probably isn't the best factoring, we could also calculate these parameters within the methods and just pass in a ```TableExpressionBase``` but there are properties of this construction that would be calculated in three different places in that factoring.

anpete replied:
Move the alias construction into the main method and pass the alias to the func.

maumar commented on src/EntityFramework.Relational/Query/Expressions/RawSqlDerivedTableExpression.cs:
```diff
@@ -0,0 +1,46 @@
+
+        public override string ToString()
+        {
+            return RawSql + " " + Alias;
```
add parentheses around RawSql for more readability

mikary replied:
That's not a pattern I've seen much before, how does it aid readability?

anpete commented on src/EntityFramework.Relational/RelationalDbSet`.cs:
```diff
@@ -0,0 +1,36 @@
+                    () => ((IAccessor<IServiceProvider>)dbSet).Service.GetRequiredService<EntityQueryProvider>());
+        }
+
+        public virtual IQueryable<TEntity> FromSql([NotNull]string query)
```
Use extension method rather than custom DbSet.

mikary replied:
Any suggestions for making sure this doesn't leak outside of the relational context?

anpete replied:
It only "lights up" if you are referencing relational.

anpete commented on test/EntityFramework.Relational.FunctionalTests/QueryRelationalTest.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Microsoft.Data.Entity.Relational.FunctionalTests
+{
+    public class QueryRelationalTest<TFixture> : QueryTestBase<TFixture>
```
Don't inherit here as the base tests are going to be run twice. Refactor the asserts into a test helper if required.
Make this test base class abstract.

anpete commented on test/EntityFramework.Relational.FunctionalTests/AsyncQueryRelationalTest.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace Microsoft.Data.Entity.Relational.FunctionalTests
+{
+    public class AsyncQueryRelationalTest<TFixture> : AsyncQueryTestBase<TFixture>
```
Don't inherit here as the base tests are going to be run twice. Refactor the asserts into a test helper if required.
Make this test base class abstract.

anpete commented on src/EntityFramework.Core/Query/IEntityQueryable.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Microsoft.Data.Entity.Query
+{
+    public interface IEntityQueryable : IMetadata, IQueryable
```
What is the purpose of this interface?

mikary replied:
It is intended to simplify the casting logic in the Expression Tree Visitors.

anpete commented on src/EntityFramework.Core/Query/ExpressionTreeVisitors/EntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -19,10 +19,11 @@ protected EntityQueryableExpressionTreeVisitor([NotNull] EntityQueryModelVisitor
         {
-            if (constantExpression.Type.GetTypeInfo().IsGenericType
-                && constantExpression.Type.GetGenericTypeDefinition() == typeof(EntityQueryable<>))
+            var entityQueryable = constantExpression.Value as IEntityQueryable;
```
This need to change?

mikary replied:
Need is a strong word, but this logic is common between both visitors and the reflection work seems a bit heavyweight for identifying a type and pulling out non-generic property values

anpete replied:
I would prefer to leave this code as it - unless it is needed for this change.

anpete commented on src/EntityFramework.Relational/Query/Expressions/RawSqlDerivedTableExpression.cs:
```diff
@@ -0,0 +1,46 @@
+    public class RawSqlDerivedTableExpression : TableExpressionBase
+    {
+        public RawSqlDerivedTableExpression(
+            [NotNull] string rawSql,
```
"sql"

anpete commented on src/EntityFramework.Relational/Query/Expressions/TableExpression.cs:
```diff
@@ -22,7 +22,6 @@ public class TableExpression : TableExpressionBase
                 Check.NotEmpty(alias, nameof(alias)))
         {
             Check.NotEmpty(table, nameof(table));
-            Check.NotNull(querySource, nameof(querySource));
```
Good catch.

anpete commented on src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs:
```diff
@@ -657,5 +668,12 @@ protected virtual string DelimitIdentifier([NotNull] string identifier)
             return "\"" + identifier + "\"";
         }
+
+        protected virtual string DelimitSubQuery([NotNull] string query)
```
If this is not overridden, I would just inline it.

anpete commented on test/EntityFramework.SqlServer.FunctionalTests/QuerySqlServerTest.cs:
```diff
@@ -1887,6 +1887,54 @@ public override void Where_chain()
+
+            Assert.Equal(
+                @"SELECT [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[CustomerID], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
+FROM (SELECT * FROM Customers WHERE Customers.ContactName LIKE '%o%') AS [c]
```
Start the subquery on a new line and indented - see Take_Distinct_Count for example.

mikary commented:
More updates based on PR input

divega commented on test/EntityFramework.SqlServer.FunctionalTests/FromSqlQuerySqlServerTest.cs:
```diff
@@ -0,0 +1,78 @@
+
+SELECT [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[CustomerID], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
+FROM (
+    SELECT * FROM Customers WHERE Customers.City = 'Seattle'
```
If the raw SQL query contains multiple lines, will they all be indented the same?

mikary replied:
There isn't any logic in place to do this right now, did we want to try?

anpete replied:
Should be easy if you use a StringReader to read each line and append to the output.

mikary replied:
Would we need to worry about literal line-breaks in quoted strings etc.?

anpete replied:
I don't think you can do that - at least in TSQL.

divega replied:
Maybe it is not to bad, but the alternative is to not worry about it at all and not try to indent raw SQL.

divega commented on src/EntityFramework.Relational/RelationalDbSetExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public static class RelationalDbSetExtensions
+    {
+        public static IQueryable<TEntity> FromSql<TEntity>([NotNull]this DbSet<TEntity> dbSet, [NotNull]string query) where TEntity : class
```
What is the parameter story?

mikary replied:
I was just looking to get the basic DbSet pipeline in place with this PR, we still have parameters, best-effort materialization(off of the Database) and possibly dynamic types.

anpete replied:
Yep, let's do that separately.

anpete replied:
Reformat this line, move constraint to new line.

divega replied:
What if the return type of this method was not just IQueryable, but something that still denotes it is a query root, e.g. DbSet? I thought about it because I was wondering how to enable the chaining we talked about today, e.g. of different provider specific raw query methods, but I think beyond that it would make sense since the method just adds information to the DbSet. Thoughts?

mikary replied:
We could have some kind of ```DbSet<T>``` wrapper or builder returned instead of IQueryable.

I'm currently working with a change where I added ```IAccessor<EntityQueryable>``` to ```DbSet<T>```, but the extension method clones the query before annotating so it still won't chain like that yet

divega commented:
@mikary @anpete @maumar Is there any more reviewing happening on this or are we good with it? I think we should figure out handling parameters (unless I am missing something) but that could go in a different commit.

maumar commented on src/EntityFramework.Relational/Query/Expressions/RawSqlDerivedTableExpression.cs:
```diff
@@ -0,0 +1,46 @@
+
+        public override string ToString()
+        {
+            return Sql + " " + Alias;
```
"(" + Sql + ") AS " + Alias?

mikary replied:
That doesn't seem very DRY to me, since this would be guessing at the interpretation of the SQL generator that consumes this object.

The current implementation is consistent with the other TableExpressions, but we could revisit this if it's important / useful enough.

anpete replied:
This is just for debugging and so do whatever makes it clear.

mikary replied:
Added #1795 to track this

mikary commented:
Updated based on feedback

anpete commented on src/EntityFramework.Core/Query/EntityQueryable`.cs:
```diff
@@ -28,5 +34,47 @@ IAsyncEnumerator<TResult> IAsyncEnumerable<TResult>.GetEnumerator()
+
+        public virtual Annotation AddAnnotation([NotNull] string annotationName, [NotNull] string value)
+        {
+            Check.NotNull(annotationName, nameof(annotationName));
```
NotEmpty

mikary replied:
That's a stronger constraint than we have on ```Annotatable``` itself, do we want to update that as well?

anpete replied:
Yes, seems like an oversight.

anpete commented on src/EntityFramework.Relational/Query/ExpressionTreeVisitors/RelationalEntityQueryableExpressionTreeVisitor.cs:
```diff
@@ -68,25 +71,65 @@ protected override Expression VisitMethodCallExpression([NotNull] MethodCallExpr
         {
             Check.NotNull(elementType, nameof(elementType));
 
+            return VisitSelectExpression(elementType,
```
Calling this "Visit" is confusing. How about "CreateQuery"?

anpete commented on src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs:
```diff
@@ -177,6 +178,31 @@ public virtual Expression VisitSelectExpression(SelectExpression selectExpressio
+        {
+            Check.NotNull(rawSqlDerivedTableExpression, nameof(rawSqlDerivedTableExpression));
+
+            _sql.Append("(")
```
Just AppendLine?

anpete commented on src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs:
```diff
@@ -177,6 +178,31 @@ public virtual Expression VisitSelectExpression(SelectExpression selectExpressio
+
+            using (_sql.Indent())
+            {
+                var reader = new StringReader(rawSqlDerivedTableExpression.Sql);
```
Disposable

anpete commented on src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs:
```diff
@@ -177,6 +178,31 @@ public virtual Expression VisitSelectExpression(SelectExpression selectExpressio
+                var reader = new StringReader(rawSqlDerivedTableExpression.Sql);
+
+                var line = reader.ReadLine();
+                while (line != null)
```
string line;
while ((line = reader.ReadLine()) != null) {

mikary replied:
I think this would harm readability

anpete replied:
This is a well known pattern for avoiding the duplicate calls to ReadLine.

anpete commented on src/EntityFramework.Relational/RelationalDbSetExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+            Check.NotNull(dbSet, nameof(dbSet));
+            Check.NotNull(query, nameof(query));
+
+            if (dbSet.Context.Database as RelationalDatabase == null)
```
Remove this check (and the Context property). If we are in this method then we can assume relational is in use.

mikary replied:
Without this, someone can include relational and make a call on a non-relational context (i.e. InMemoryStore) and it will silently ignore the custom SQL.

anpete replied:
Seems unlikely. Either way, the way this is checked seems hacky. I don't think we can guarantee that Database will always be a RelationalDatabase. The type of the concrete Database class only seems coincidentally related to whether or not the underlying query provider understands the annotation.

anpete commented on src/EntityFramework.Relational/RelationalDbSetExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+                throw new InvalidOperationException(Strings.RelationalNotInUse);
+            }
+
+            var queryProvider = ((IAccessor<IServiceProvider>)dbSet).Service.GetRequiredService<EntityQueryProvider>();
```
Rather than do this, explicitly implement IAnnotatable on DbSet<> and then provide an impl in InternalDbSet. The impl can just delegate to _entityQueryable.

mikary replied:
I'm not sure I understand what you mean. I'm getting the provider so that I can create a query and creating a new EntityQueryable to make sure we aren't persisting any annotations beyond this call.

anpete replied:
We always have a unique EntityQueryable instance per DbSet.

anpete commented on test/EntityFramework.Core.Tests/TestHelpers.cs:
```diff
@@ -174,5 +177,39 @@ public ModelBuilder CreateConventionBuilder(Model model = null)
             return entry;
         }
+
+        public static int AssertResults<T>(
```
DRY this up everywhere. Rename to AssertQuery.

mikary replied:
Do we want all of the ```AssertQuery``` (sync and async) moved in here?

anpete replied:
Only where it makes it DRY'er

anpete commented on test/EntityFramework.Relational.FunctionalTests/FromSqlQueryTestBase.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [Fact]
+        public virtual void From_sql_queryable_where_simple_closure_via_query_cache()
```
Test the same query but with different FromSql.

mikary replied:
I thought that case was covered by ```From_sql_queryable_cached_by_query```
is this test just redundant?

anpete replied:
Oh, I see that now. Just wondering how it works because it seems like there should be a single cache entry that is bound to the first SQL statement.

mikary commented:
Rebased, squashed, updated, and ready for another pass.
Of note:
* Simplified extension method
* DataStore now has an ```InSupported(IAnnotation)``` method that is checked when the query is compiled
* Simplified Query Tree Visitor (added ```IAnnotatable``` parameter to constant expression and simplified implementation logic)

mikary commented:
After talk with @anpete, looking at using QueryAnnotations to store the FromSql string and parameters.

divega commented:
@mikary @anpete Are we going to check in what we have and then do that or just keep iterating on this PR?

anpete commented:
This is mostly a rewrite so a new PR.

mikary commented:
Closing PR, will make a new one for the next revision.
